### PR TITLE
implement quadratic CVaR

### PR DESCRIPTION
**Pull request by masanorihirano**

### CODE REVIEW

File: pfhedge/nn/functional.py [hunk 1]
```diff
@@ -321,6 +322,75 @@ def value_at_risk(input: Tensor, p: float, dim: Optional[int] = None) -> Tensor:
...
+            if not dim:
+                raise AssertionError("Unknown Error (dim is modified inappropriately.)")
+            return fn.relu(-_omega.unsqueeze(dim=dim) - input).mean(dim=dim)
+
+    else:
+        base = input.mean()
+        input -= base
+
+        def fn_target(_omega: Tensor) -> Tensor:
+            return fn.relu(-_omega - input).mean()
```
Comment (imos) [inline pfhedge/nn/functional.py:370]: dim == 0の扱いが怪しい。おそらく整理するとこんな感じで書ける。unsqueezeするのではなく、本当に不要になったときにmeanなりで次元を潰した方が取り回しやすいはず。

```suggestion
    output_target = torch.as_tensor(1 / (2 * lam))
    base = input.mean(dim=dim, keepdim=True)
    input = input - base

    def fn_target(omega: Tensor) -> Tensor:
        return fn.relu(-omega - input).mean(dim=dim)
```

File: pfhedge/nn/functional.py [hunk 2]
```diff
@@ -321,6 +322,75 @@ def value_at_risk(input: Tensor, p: float, dim: Optional[int] = None) -> Tensor:
...
+
+    Examples:
+        >>> from pfhedge.nn.functional import quadratic_cvar
+        >>>
+        >>> input = -torch.arange(10.0)
+        >>> input
+        tensor([-0., -1., -2., -3., -4., -5., -6., -7., -8., -9.])
+        >>> quadratic_cvar(input, 2.0)
+        tensor(7.9750)
+    """  # NOQA
```
Comment (imos) [inline pfhedge/nn/functional.py:354]: dim is None は別処理した方が確実かつ読みやすい

```suggestion
    """  # NOQA
    if dim is None:
        return quadratic_cvar(input.flatten(), lam, 0)
```

File: pfhedge/nn/functional.py [hunk 3]
```diff
@@ -321,6 +322,75 @@ def value_at_risk(input: Tensor, p: float, dim: Optional[int] = None) -> Tensor:
...
+        precision=precision,
+    )
+    if dim:
+        return (
+            omega
+            + lam * fn.relu(-omega.unsqueeze(dim=dim) - input).square().mean(dim=dim)
+            - base
+        )
+    else:
+        return omega + lam * fn.relu(-omega - input).square().mean() - base
```
Comment (imos) [inline pfhedge/nn/functional.py:391]: dim is None のケースがなくなればシンプルに書けるはず。

```suggestion
    return omega + lam * fn.relu(-omega.unsqueeze(dim=dim) - input).square().mean(dim=dim) - base
```

**Review by imos — CHANGES_REQUESTED** [review]

**Comment by masanorihirano** [comment]
dim=Noneとdim=0の挙動は異なるはずだったので，追加コミットで修正します．
 - dim=None : 次元のないスカラーを返す
 - dim=0 : 0次元目でつぶす

File: pfhedge/nn/functional.py [hunk 4]
```diff
@@ -321,6 +322,64 @@ def value_at_risk(input: Tensor, p: float, dim: Optional[int] = None) -> Tensor:
...
+    for :math:`\lambda\geq1`.
+
+    References:
+        - Buehler, Hans, Statistical Hedging (March 1, 2019). Available at SSRN: http://dx.doi.org/10.2139/ssrn.2913250
+          (See Conclusion.)
+
+    Args:
+        input (torch.Tensor): The input tensor.
+        lam (float): :math:`lambda`.
+        dim (int, optional): The dimension to sort along.
```
Comment (imos) [inline pfhedge/nn/functional.py:341]:
```suggestion
        input (torch.Tensor): The input tensor.
        lam (float): The :math:`lambda` parameter, representing the weight given to the tail losses.
        dim (int, optional): The dimension to sort along. If None, the tensor is flattened.
```

File: pfhedge/nn/functional.py [hunk 5]
```diff
@@ -321,6 +322,64 @@ def value_at_risk(input: Tensor, p: float, dim: Optional[int] = None) -> Tensor:
...
+
+    Returns:
+        torch.Tensor
```
Comment (imos) [inline pfhedge/nn/functional.py:344]:
```suggestion
        torch.Tensor: The Quadratic CVaR of the input tensor.
```

File: pfhedge/nn/functional.py [hunk 6]
```diff
@@ -321,6 +322,64 @@ def value_at_risk(input: Tensor, p: float, dim: Optional[int] = None) -> Tensor:
...
+        >>>
+        >>> input = -torch.arange(10.0)
+        >>> input
+        tensor([-0., -1., -2., -3., -4., -5., -6., -7., -8., -9.])
+        >>> quadratic_cvar(input, 2.0)
+        tensor(7.9750)
+    """  # NOQA
+    if dim is None:
+        return quadratic_cvar(input.flatten(), lam, 0)
+    output_target = torch.as_tensor(1 / (2 * lam))
```
Comment (imos) [inline pfhedge/nn/functional.py:357]: 意味が分かれるので分ける

```suggestion
    if dim is None:
        return quadratic_cvar(input.flatten(), lam, 0)

    output_target = torch.as_tensor(1 / (2 * lam))
```

**Review by imos — APPROVED** [review]
